tools/btmidi-server.c

```diff
@@ -0,0 +1,927 @@
+						BT_ATT_PERM_READ | BT_ATT_PERM_WRITE,
+						BT_GATT_CHRC_PROP_WRITE_WITHOUT_RESP | BT_GATT_CHRC_PROP_READ | BT_GATT_CHRC_PROP_NOTIFY,
+						NULL, midi_write_cb, server);
+	server->midi_io_handle = gatt_db_attribute_get_handle(server->midi_io);
+
+	/* Add MIDI CCC */
+	bt_uuid16_create(&uuid, GATT_CLIENT_CHARAC_CFG_UUID);
+	gatt_db_service_add_descriptor(service, &uuid,
+					BT_ATT_PERM_READ | BT_ATT_PERM_WRITE,
+					midi_ccc_read_cb, midi_ccc_write_cb, server);
+
+	/* Activate Service */
+	gatt_db_service_set_active(service, true);
+
+	return 0;
+
+_err_midi:
+	midi_read_free(&server->midi_in);
+
+_err_port:
+	snd_seq_delete_simple_port(server->seq_handle, server->seq_port_id);
+
+_err_handle:
+	snd_seq_close(server->seq_handle);
+	server->seq_handle = NULL;
+
+	return -1;
+}
+
+static void add_adv_callback(uint8_t status, uint16_t length,
+							const void *param, void *user_data)
+{
+// 	struct btd_adv_client *client = user_data;
+	const struct mgmt_rp_add_advertising *rp = param;
+
+	if (status)
+		goto done;
+
+	if (!param || length < sizeof(*rp)) {
+		status = MGMT_STATUS_FAILED;
+		goto done;
+	}
+done:
+	printf("Blz\n");
+}
+
+static int advertise(int dev_id)
+{
+	struct mgmt *mgmt_if = NULL;
+	struct bt_ad *data;
+	uint8_t *adv_data;
+	size_t adv_data_len;
+	struct mgmt_cp_add_advertising *cp;
+	uint8_t param_len;
+	bt_uuid_t uuid;
+	
+	bt_string_to_uuid(&uuid, MIDI_UUID);
+
+	mgmt_if = mgmt_new_default();
+	if (!mgmt_if) {
+		fprintf(stderr, "Failed to access management interface\n");
+		return -1;
+	}
+
+	data = bt_ad_new();
+	if (!data) {
+		fprintf(stderr, "Error creating adverting data\n");
+		return -1;
+	}
+
+	if (!bt_ad_add_service_uuid(data, &uuid)) {
+		fprintf(stderr, "Error adding service UUID\n");
+		return -1;
+	}
+
+	adv_data = bt_ad_generate(data, &adv_data_len);
+	if (!adv_data) {
+		fprintf(stderr, "Error generating advertising data\n");
+		return -1;
+	}
+
+	param_len = sizeof(struct mgmt_cp_add_advertising) + adv_data_len;
+	cp = malloc0(param_len);
+	if (!cp) {
+		fprintf(stderr, "Error allocating memory to advertising params\n");
+		return -1;
+	}
+
+	cp->flags = htobl(MGMT_ADV_FLAG_CONNECTABLE | MGMT_ADV_FLAG_DISCOV);
+	cp->instance = 0;
+	cp->adv_data_len = adv_data_len;
+	memcpy(cp->data, adv_data, adv_data_len);
+
+	free(adv_data);
+
+	if (!mgmt_send(mgmt_if, MGMT_OP_ADD_ADVERTISING, dev_id, param_len,
+				cp, add_adv_callback, NULL, NULL)) {
+		fprintf(stderr, "Failed to add Advertising Data");
+		free(cp);
+		return -1;
+	}
+
+	free(cp);
+	bt_ad_unref(data);
+	mgmt_unref(mgmt_if);
+
+	return 0;
+}
+
+static int l2cap_le_att_listen_and_accept(bdaddr_t *src, int sec,
+							uint8_t src_type)
+{
+	int sk, nsk;
+	struct sockaddr_l2 srcaddr, addr;
+	socklen_t optlen;
+	struct bt_security btsec;
+	char ba[18];
+	int dev_id;
+	
+	sk = socket(PF_BLUETOOTH, SOCK_SEQPACKET, BTPROTO_L2CAP);
+	if (sk < 0) {
+		perror("Failed to create L2CAP socket");
+		return -1;
+	}
+
+	dev_id = hci_get_route(src);
```

Comment:
It's just that you don't need to do it since it's already been verified in the main function, right?